File: lorcon/examples/beacon_flood_raw.py
#!/usr/bin/env python
#
# beacon_flood_raw.py - Simple IEEE 802.11
#	beacon flooder using pylorcon2's raw
#	sending capabilities.
#
# brad.antoniewicz@foundstone.com
#
#

import getopt
import sys
import string

from time import sleep

import PyLorcon2

def usage():
	print "\t-i <int> \tInterface"
	print "\t-c <channel> \tChannel"
	print "\nExample: "
	print "\t",sys.argv[0],"-i wlan0 -c 11\n"


'''
	main
'''

print sys.argv[0]," - Simple 802.11 beacon flooder"
print "-----------------------------------------------------\n" 

# Beacon interval 
interval = 100

# Raw packet bytes (from capture_example.c included within LORCON)
packet = 	"\x80\x00\xff\xff\xff\xff\xff\xff" \
		"\xff\xff\x00\x0f\x66\xe3\xe4\x03" \
		"\x00\x0f\x66\xe3\xe4\x03\x00\x00" \
		"\xff\xff\xff\xff\xff\xff\xff\xff" \
		"\x64\x00\x11\x00\x00\x0f\x73\x6f" \
		"\x6d\x65\x74\x68\x69\x6e\x67\x63" \
		"\x6c\x65\x76\x65\x72\x01\x08\x82" \
		"\x84\x8b\x96\x24\x30\x48\x6c\x03" \
		"\x01\x01\x05\x04\x00\x01\x00\x00" \
		"\x2a\x01\x05\x2f\x01\x05\x32\x04" \
		"\x0c\x12\x18\x60\xdd\x05\x00\x10" \
		"\x18\x01\x01\xdd\x16\x00\x50\xf2" \
		"\x01\x01\x00\x00\x50\xf2\x02\x01" \
		"\x00\x00\x50\xf2\x02\x01\x00\x00" \
		"\x50\xf2\x02"

interface = channel = None 

'''
        This handles all of the command line arguments
'''


try:
	opts, args = getopt.getopt(sys.argv[1:], "i:c:h",[])
except getopt.GetoptError:
	usage()
	sys.exit(-1)	

for o,a in opts:
	if o == "-i":
		interface = a
	if o == "-c":
		channel = string.atoi(a)
	if o == "-h":
		usage()

if ( (interface is None) or (channel is None) ):
	print "ERROR: Interface or channel not set (see -h for more info)"
	sys.exit(-1)

print "[+] Using interface",interface

'''
	The following is all of the standard interface, driver, and context setup
'''

# Automatically determine the driver of the interface

try:
	driver, description = PyLorcon2.auto_driver(interface)
	if driver is not None:
		print "[+]\t Driver:",driver
except:
	print "[!] Could not determine the driver for",interface
	sys.exit(-1)

# Create LORCON context
try:
	ctx = PyLorcon2.Context(interface)
except:
	print "[!]\t Failed to create context"
	sys.exit(-1)

# Create Monitor Mode Interface

try:
	ctx.open_injmon()
	vap = ctx.get_vap()
	if vap is not None:	
		print "[+]\t Monitor Mode VAP:",vap
except:
	print "[!]\t Could not create Monitor Mode interface!"
	sys.exit(-1)

# Set the channel we'll be injecting on

try:
	ctx.set_channel(channel)
	print "[+]\t Using channel:",channel,"\n"
except:
	print "[!]\t Could not set channel!"
	sys.exit(-1)	

'''
	The following is the packet creation and sending code
'''

# Keep sending frames until interrupted
count = 0
while 1:

	try:
		ctx.send_bytes(packet)
		sleep(interval/1000.0)
		print "[+] Sent",count,"frames, Hit CTRL + C to stop...\r",
		sys.stdout.flush()
				
	except:
		print "\n[!] Exiting..."
		sys.exit(0)
	
	count+=1


'''
	The following is all of the standard cleanup stuff
'''

# Close the interface
ctx.close()

File: lorcon/examples/beacon_flood_raw.c
/*
	beacon_flood_raw.c 
	by brad.antoniewicz@foundstone.com	

	simple IEEE 802.11 beacon flooder using LORCON2's 
	raw sending capabilities

*/

#include <stdio.h>
#include <getopt.h>

#include <lorcon2/lorcon.h> // For LORCON 

void usage(char *argv[]) {
	printf("\t-i <int> \tInterface\n");
	printf("\t-c <channel>\tChannel\n");
	printf("\nExample:\n");
	printf("\t%s -i wlan0 -c 1\n\n",argv[0]);
}
int main(int argc, char *argv[]) {

	char *interface = NULL, *ssid = NULL;
	int c, channel;
	unsigned int count=0;

	lorcon_driver_t *drvlist, *driver; // Needed to set up interface/context
	lorcon_t *context; // LORCON context

	// Beacon Interval
        int interval = 100;

	// Raw packet bytes (from capture_example.c included within LORCON)
	unsigned char packet[115] = {
        0x80, 0x00, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, // dur ffff
        0xff, 0xff, 0x00, 0x0f, 0x66, 0xe3, 0xe4, 0x03,
        0x00, 0x0f, 0x66, 0xe3, 0xe4, 0x03, 0x00, 0x00, // 0x0000 - seq no.
        0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, // BSS timestamp
        0x64, 0x00, 0x11, 0x00, 0x00, 0x0f, 0x73, 0x6f,
        0x6d, 0x65, 0x74, 0x68, 0x69, 0x6e, 0x67, 0x63,
        0x6c, 0x65, 0x76, 0x65, 0x72, 0x01, 0x08, 0x82,
        0x84, 0x8b, 0x96, 0x24, 0x30, 0x48, 0x6c, 0x03,
        0x01, 0x01, 0x05, 0x04, 0x00, 0x01, 0x00, 0x00,
        0x2a, 0x01, 0x05, 0x2f, 0x01, 0x05, 0x32, 0x04,
        0x0c, 0x12, 0x18, 0x60, 0xdd, 0x05, 0x00, 0x10,
        0x18, 0x01, 0x01, 0xdd, 0x16, 0x00, 0x50, 0xf2,
        0x01, 0x01, 0x00, 0x00, 0x50, 0xf2, 0x02, 0x01,
        0x00, 0x00, 0x50, 0xf2, 0x02, 0x01, 0x00, 0x00,
        0x50, 0xf2, 0x02};



	printf ("%s - Simple 802.11 beacon flooder\n", argv[0]);
	printf ("-----------------------------------------------------\n\n");

	/* 
		This handles all of the command line arguments
	*/
	
	while ((c = getopt(argc, argv, "i:s:hc:")) != EOF) {
		switch (c) {
			case 'i': 
				interface = strdup(optarg);
				break;
			case 'c':
				channel = atoi(optarg);
				break;
			case 'h':
				usage(argv);
				break;
			default:
				usage(argv);
				break;
			}
	}

	if ( interface == NULL  ) { 
		printf ("ERROR: Interface not set (see -h for more info)\n");
		return -1;
	}

	printf("[+] Using interface %s\n",interface);
	
	/*	
	 	The following is all of the standard interface, driver, and context setup
	*/

	// Automatically determine the driver of the interface
	
	if ( (driver = lorcon_auto_driver(interface)) == NULL) {
		printf("[!] Could not determine the driver for %s\n",interface);
		return -1;
	} else {
		printf("[+]\t Driver: %s\n",driver->name);
	}

	// Create LORCON context
        if ((context = lorcon_create(interface, driver)) == NULL) {
                printf("[!]\t Failed to create context");
               	return -1; 
        }

	// Create Monitor Mode Interface
	if (lorcon_open_injmon(context) < 0) {
		printf("[!]\t Could not create Monitor Mode interface!\n");
		return -1;
	} else {
		printf("[+]\t Monitor Mode VAP: %s\n",lorcon_get_vap(context));
		lorcon_free_driver_list(driver);
	}

	// Set the channel we'll be injecting on
	lorcon_set_channel(context, channel);
	printf("[+]\t Using channel: %d\n\n",channel);

	/* 
		The following is the packet creation and sending code
	*/

	// Keep sending frames until interrupted
	while(1) {

		// Send and exit if error
		if ( lorcon_send_bytes(context, sizeof(packet), packet) < 0 ) 
			return -1;

               // Wait interval before next beacon
                usleep(interval * 1000);

		// Print nice and pretty
		printf("\033[K\r");
		printf("[+] Sent %d frames, Hit CTRL + C to stop...", count);
		fflush(stdout);
		count++;

	}

	/* 
	 	The following is all of the standard cleanup stuff
	*/

	// Close the interface
	lorcon_close(context);

	// Free the LORCON Context
	lorcon_free(context);	
	
	return 0;
}

